test/flush_worker_test.cc

```diff
@@ -0,0 +1,161 @@
+using ::testing::AtLeast;
+using ::testing::NiceMock;
+using ::testing::ReturnRef;
+using ::testing::StrictMock;
+using ::testing::Test;
+
+class FlushWorkerTest : public Test {
+public:
+  FlushWorkerTest()
+      : thread_factory_(Envoy::Thread::threadFactoryForTest()),
+        dispatcher_(new NiceMock<Envoy::Event::MockDispatcher>()) {
+    Envoy::Random::RandomGeneratorImpl rand;
+    NiceMock<Envoy::LocalInfo::MockLocalInfo> local_info;
+    NiceMock<Envoy::ProtobufMessage::MockValidationVisitor> validation_visitor;
+    loader_ = std::make_unique<Envoy::Runtime::ScopedLoaderSingleton>(
+        Envoy::Runtime::LoaderPtr{new Envoy::Runtime::LoaderImpl(
+            *dispatcher_, tls_, {}, local_info, store_, rand, validation_visitor, mock_api_)});
+    sink_ = new StrictMock<Envoy::Stats::MockSink>();
+    stats_sinks_.emplace_back(sink_);
+
+    EXPECT_CALL(mock_api_, threadFactory()).WillRepeatedly(ReturnRef(thread_factory_));
+    // mock_api_ takes the ownership of dispatcher_.
+    EXPECT_CALL(mock_api_, allocateDispatcher_(_, _)).WillOnce(Return(dispatcher_));
+  }
+
+  // Set up emulating timer firing and corresponding expectations and behaviors.
+  void setupDispatcherTimerEmulation() {
+    timer_ = new NiceMock<Envoy::Event::MockTimer>();
+    EXPECT_CALL(*dispatcher_, createTimer_(_)).WillOnce(Invoke([&](Envoy::Event::TimerCb cb) {
+      timer_cb_ = std::move(cb);
+      return timer_;
+    }));
+    EXPECT_CALL(*timer_, enableTimer(_, _))
+        .WillRepeatedly(Invoke([&](const std::chrono::microseconds,
+                                   const Envoy::ScopeTrackedObject*) { timer_set_ = true; }));
+    EXPECT_CALL(*timer_, disableTimer()).WillOnce(Invoke([&]() { timer_set_ = false; }));
+    EXPECT_CALL(*dispatcher_, exit()).WillOnce(Invoke([&]() {
+      auto guard = std::make_unique<Envoy::Thread::LockGuard>(lock_);
+      stopped_ = true;
+    }));
+  }
+
+  // Set up expected behaviors when run() is called on dispatcher_.
+  void expectDispatcherRun() {
+    EXPECT_CALL(*dispatcher_, run(_))
+        .WillOnce(Invoke([&](Envoy::Event::DispatcherImpl::RunType type) {
+          // The first dispatcher run is in WorkerImpl::start().
+          ASSERT_EQ(Envoy::Event::DispatcherImpl::RunType::NonBlock, type);
+        }))
+        .WillOnce(Invoke([&](Envoy::Event::DispatcherImpl::RunType type) {
+          // The second dispatcher run is in FlushWorkerImpl::work().
+          ASSERT_EQ(Envoy::Event::DispatcherImpl::RunType::RunUntilExit, type);
+          simulateTimerLoop();
+        }));
+  }
+
+  // Simulate the periodical timer which keeps running until dispatcher->exit()
+  // is called from another thread. The std::promise signal_dispatcher_to_exit_
+  // is used here to guarantee the while loop runs for at least 100 times before dispatcher->exit()
+  // is called.
+  void simulateTimerLoop() {
+    int i = 0;
+    while (true) {
+      i++;
+      // At least run the while loop for 100 times.
+      if (i == 100) {
```

Comment:
It certainly does raise a question of why is this 100, so it is a good idea to put a comment here explaining it. Unfortunately the current comment doesn't help and goes against the style guide as it just describes the code:

https://google.github.io/styleguide/cppguide.html#Implementation_Comment_Donts

Can we try to improve this comment? Also I think @oschaaf had a suggestion to replace the literal 100 with a well named constant which could improve this further.

Reply:
Thanks Jakub. Replaced the 100 with a constant and updated the comment clarifying it does not to be 100.